### PR TITLE
addpkg(x11/xfce4-cpugraph-plugin): 1.2.10

### PR DESCRIPTION
**Pull request by knyipab**
On my phone with snapdragon 8 gen 2, core 0 and core 1 are usually not shown in `top -H -b -q -o TID,%CPU,CPU -s 2 -n 1` for unknown reasons, it counts towards other cores. It's more of `top` issue not my code. Still it is useful to show the area graph. 

@hansm629, please check and any feedback.

### CODE REVIEW

**Comment by hansm629** [comment]
@knyipab
Tested on `Exynos 2400`.

- 3.2GHz Cortex-X4 1core
- 2.9GHz Cortex-A720 2 core
- 2.6GHz Cortex-A720 3 core
- 1.96GHz Cortex-A520 4 core
- L3 cache 8MB + SLC 8MB

`ARM64 10 core` specifications.





When rendering `4K@30fps HEVC video` in `Shotcut`
All 10 core CPUs are being monitored!

Excellent!

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data.patch
Comment (licy183) [inline x11-packages/xfce4-cpugraph-plugin/0001-cpu-data.patch:57]: This actually only works on Android 9.0 or later, but I don't think there are many users running X11 programs on Android 7 and Android 8, so I think this is acceptable here. 

PS: I'm just commenting here in case other devs port similar logic to some other frequently used tool (e.g. `htop`).

https://android.googlesource.com/platform/external/toybox/+/refs/tags/android-7.0.0_r36/toys/posix/ps.c#124
```
  help
    usage: top [-m] [ -d seconds ] [ -n iterations ]
    Show process activity in real time.
    -k	Fallback sort FIELDS (default -S,-%CPU,-ETIME,-PID)
    -o	Show FIELDS (def PID,USER,PR,NI,VIRT,RES,SHR,S,%CPU,%MEM,TIME+,CMDLINE)
    -s	Sort by field number (1-X, default 9)
```

https://android.googlesource.com/platform/external/toybox/+/refs/tags/android-8.0.0_r51/toys/posix/ps.c
```
  help
    usage: top [-H] [-k FIELD,] [-o FIELD,] [-s SORT]
    Show process activity in real time.
    -H	Show threads
    -k	Fallback sort FIELDS (default -S,-%CPU,-ETIME,-PID)
    -o	Show FIELDS (def PID,USER,PR,NI,VIRT,RES,SHR,S,%CPU,%MEM,TIME+,CMDLINE)
    -O	Add FIELDS (replacing PR,NI,VIRT,RES,SHR,S from default)
    -s	Sort by field number (1-X, default 9)
```

https://android.googlesource.com/platform/external/toybox/+/refs/tags/android-9.0.0_r61/toys/posix/ps.c

```
  help
    usage: top [-Hbq] [-k FIELD,] [-o FIELD,] [-s SORT] [-n NUMBER] [-m LINES] [-d SECONDS] [-p PID,] [-u USER,]
    Show process activity in real time.
    -H	Show threads
    -k	Fallback sort FIELDS (default -S,-%CPU,-ETIME,-PID)
    -o	Show FIELDS (def PID,USER,PR,NI,VIRT,RES,SHR,S,%CPU,%MEM,TIME+,CMDLINE)
    -O	Add FIELDS (replacing PR,NI,VIRT,RES,SHR,S from default)
    -s	Sort by field number (1-X, default 9)
    -b	Batch mode (no tty)
    -d	Delay SECONDS between each cycle (default 3)
    -m	Maximum number of tasks to show
    -n	Exit after NUMBER iterations
    -p	Show these PIDs
    -u	Show these USERs
    -q	Quiet (no header lines)
    Cursor LEFT/RIGHT to change sort, UP/DOWN move list, space to force
    update, R to reverse sort, Q to exit.
```

Comment (licy183) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data.patch:57]: `/proc/stat` is prohibited from reading since Android 9.0. It would be better to dynamically check the api level using `android_get_device_api_level` and select which method to use. If >= 28, use the htop method, otherwise use `/proc/stat` method.

**Comment by knyipab** [comment]
I rewrote to use thread stat aggregation instead of `top`. It should be more efficient and accurate. 

I will probably write another patch for `htop` sometime later. 

Again, couldn't use first two cores on my device (perhaps due to phone performance policy). @hansm629 see if you are interested in trying this version (perhaps along with the most recent #20393).

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data.patch
Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data.patch]: Rewrote to use `/proc/#/task/#/stat` and keep the `top` approach in commit history just in case we need in the future.

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (tstein) [inline x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:18]: `void get_...` probably isn't right. `set_sdk_ver()` if you wanna keep it void, but given how you're using it:

```
int get_sdk_ver() {
  static int sdk_ver = -1;
  if (sdk_ver == -1) {
    // do the thing
  }
  return sdk_ver;
}
```

seems better.

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (tstein) [inline x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:21]: Please fix this patch to consistently use soft or hard tabs, whatever's in the file it applies to.

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (tstein) [inline x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:56]: What does stocktake mean?

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (tstein) [inline x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:59]: Where would `HZ` have been defined?

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (tstein) [inline x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:89]: Good call on all the comments on this format string, thank you. :)

**Comment by hansm629** [comment]
@knyipab
Are you saying that 2 of the CPU cores cannot be monitored?

- Galaxy Tab S9 Ultra (Snapdragon 8 Gen2 for Galaxy : 8 cores)
- Galaxy S24+ (Exynos 2400 : 10 cores)
- Galaxy S24 Ultra (Snapdragon 8 Gen3 for Galaxy : 8 cores)

All cores were monitored.

The screenshot below is from the Galaxy Tab S9 Ultra.
When rendering 4K video with Shotcut, all 8 cores are used.

Screenshot_2024-06-29_13-31-21
Screenshot_2024-06-29_13-31-21 (copy 1)

File: x11-packages/xfce4-cpugraph-plugin/build.sh
Comment (TomJo2000) [inline x11-packages/xfce4-cpugraph-plugin/build.sh:6]: No revision if the package is just being added.

**Review by TomJo2000 — CHANGES_REQUESTED** [review]
Can't comment on the C++ but that seems well covered in the existing reviews.

File: x11-packages/xfce4-cpugraph-plugin/build.sh
Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/build.sh:6]: There are two commits (with the first one to keep the `top` implememtation in commit history, and the second one to revise to use `/proc/#/task/#/stat`). Subject to your practice, there may be no need to add revision. Thx.

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:78]: The `GHashTable` is copied and adapted casually from `xfce4-taskmanager`. I don't hold strong view on whether to use C++ STL. Would it indeed be it good do C implementation so that it can be more easily reused by other package? It need not to be glib actually. Grateful for your advance.

File: x11-packages/xfce4-cpugraph-plugin/build.sh
Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/build.sh:6]: There are two commits (with the first one to keep the `top` implememtation in commit history, and the second one to revise to use `/proc/#/task/#/stat`). Subject to your practice, there may be no need to add revision. Thx.

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:59]: That's copied from `procps` (see below). `HZ` is probably not so meaningful for Android/Termux cuz that's probably for those who build a customized kernel. Also I should use `glong` not `gulong` otherwise the subsequent negativity test won't work. Will remove `HZ` and correct later. 

```cpp
/*
 * procps_hertz_get:
 *
 *
 * Some values in /proc are expressed in units of 1/HZ seconds, where HZ
 * is the kernel clock tick rate. One of these units is called a jiffy.
 * The HZ value used in the kernel may vary according to hacker desire.
 *
 * On some architectures, the kernel provides an ELF note to indicate
 * HZ.
 *
 * Returns:
 *  The discovered or assumed hertz value
 */
PROCPS_EXPORT long procps_hertz_get(void)
{
    long hz;

#ifdef _SC_CLK_TCK
    if ((hz = sysconf(_SC_CLK_TCK)) > 0)
        return hz;
#endif
#ifdef HZ
    return(HZ);
#endif
    /* Last resort, assume 100 */
    return 100;
}
```

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:18]: Such style is copied from `xfce4-taskamanger` casually without much thought. Agree with you and will change it later

File: x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch
Comment (licy183) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:78]: Considering that other programs will reuse the same code, it is better to use C language.

Comment (knyipab) [reply x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch:78]: I got another idea (sorry for frequently making new ideas and potentially waste the review effort): since different programs may use `/proc/stat` to compute different metrics in their own ways. How about I writing a function/library that updates and writes into a compatibility file at `$PREFIX/var/proc/stat` (and perhaps when API level <28, then symlink to `/proc/stat`)? What about a package named `libandroid-procstat` that provides `update_procstat_file()`? All other packages can just call `update_procstat_file()` before reading `$PREFIX/var/proc/stat` (in place of reading `/proc/stat`). It probably requires semaphore to avoid race condition, shall I use posix semaphore or anything else?

Or alternatively, it's perhaps more natural to register the stat file path that can output geneated by c/c++ code every time read. It's somewhat like a named pipe but I am not aware of any good way to implement that.  Good to know if there is.

**Comment by knyipab** [comment]
Thanks @licy183 and @tstein for reviewing! I made a commit to refactor. It's not yet squashed, for tracking changes. 

To @licy183 after second thought, I abandon the idea of a "compatibility file" `$PREFIX/var/proc/stat` given the complexity (potentially need shared memory hash table to track thread cpu time and careful design upon termux crashes). I rather spare time for `xfce4-battery-plugin`, `pipewire` (and its subsequent applications) and Termux:X11 patches. 

To @hansm629 thanks for testing. And yes, the 8 gen 2 in my Fold 5 never utilize first two cores in Termux, and not showing up in per-thread statistics in command `top -H -O CPU`. Probably it's some Samsung performance policy under DeX or what (??), though I already set Termux battery usage to unrestricted. 

To @TomJo2000 I do not hold strong view and please advise: 
1. if the `top` implementation in the addpkg commmit should be kept or discarded
2. whether `TERMUX_PKG_REVISION=1` should be kept or dropped if there are addpkg and enhance commit
And then I will execute accordingly and squash commits (you may execute directly if you like) before merging. Thanks.

**Comment by twaik** [comment]
> I rather spare time for ... and Termux:X11 patches.

And two PRs are waiting for rebasing :)

> Probably it's some Samsung performance policy under DeX or what (??), though I already set Termux battery usage to unrestricted.

There already were problems with disabled kernels. They are being enabled on high demand, some kind of powersaving policy that you can not manage.

> whether TERMUX_PKG_REVISION=1 should be kept or dropped if there are addpkg and enhance commit

`TERMUX_PKG_REVISION` should be increased on every Github Action run if there is no version bump. That is needed to make both apt and aptly consider it as a new package version.

**Comment by knyipab** [comment]
> And two PRs are waiting for rebasing :)

Yes, I actually meant sparing time to work on that two PRs (and that's perhaps after `xfce4-battery-plugin`). 

Side topic haha: debugging Termux:X11 requires boot my PC which generates a lot for heat in my room knowing that it's summer time now :) Nowadays I like to code in Termux in DeX and perhaps NDK toolchain is the only major prerequisite left before we can bring Android Studio to Termux. In this thread, I laid out the steps to setup Android Studio that can build SDK apps and encourage this author to bring NDK toolchain to Termux cuz I don't have that expertise: https://github.com/lzhiyong/termux-ndk/issues/156. 

> `TERMUX_PKG_REVISION` should be increased on every Github Action run if there is no version bump. That is needed to make both apt and aptly consider it as a new package version.

Okay, if revision number is "per Github Action" not "per commit" then lemme drop that `TERMUX_PKG_REVISION=1`.

**Comment by twaik** [comment]
> encourage this author to bring NDK toolchain to Termux cuz I don't have that expertise: [lzhiyong/termux-ndk#156](https://github.com/lzhiyong/termux-ndk/issues/156).

What about using this? https://github.com/android-ide/aide_ndk/

**Comment by knyipab** [comment]
Seeing no reply (yet) from the author, I actually did (little) attempt to build. According to the project: 
> we don‘t need to rebuild the whole NDK, since google already built most of it. so we only need to build the llvm toolchain, then replace the llvm inside NDK.
> Building the android-ndk, please refer to [Android Clang/LLVM Toolchain Readme Doc](https://android.googlesource.com/toolchain/llvm_android/+/master/README.md)

Not sure if I did anything wrong. The source download step already takes few tens of GB and fill up my storage drive :) so I gave up. 

Android Studio on Termux is exciting. Perhaps if Termux devs find it interesting, may try.

File: x11-packages/xfce4-cpugraph-plugin/build.sh
Comment (TomJo2000) [inline x11-packages/xfce4-cpugraph-plugin/build.sh:4]: Since you wrote a pretty sizable patch for this package, please consider setting yourself as the maintainer.

**Review by TomJo2000 — APPROVED** [review]
As far as I can evaluate it, this looks fine to me.

**Review by licy183 — APPROVED** [review]
I'll merge this PR tomorrow if there is no more review. Thanks!

**Pull request merged**
